File: ios/runanywhere-sdks-latest/sdk/runanywhere-swift/Sources/RunAnywhere/Features/STT/Services/AudioCaptureManager.swift
//
//  AudioCaptureManager.swift
//  RunAnywhere SDK
//
//  Shared audio capture utility for STT features.
//  Can be used with any STT backend (ONNX, etc.)
//

import AVFoundation
import CRACommons
import Foundation

#if os(macOS)
import AudioToolbox
import CoreAudio
#endif

/// Manages audio capture from microphone for STT services.
///
/// This is a shared utility that works with any STT backend (ONNX, etc.).
/// It captures audio at 16kHz mono Int16 format, which is the standard input format
/// for speech recognition models like Whisper.
///
/// - Works on: iOS, tvOS, and macOS using AVAudioEngine
/// - NOT supported on: watchOS (AVAudioEngine inputNode tap doesn't work reliably)
///
/// ## Usage
/// ```swift
/// let capture = AudioCaptureManager()
/// let granted = await capture.requestPermission()
/// if granted {
///     try capture.startRecording { audioData in
///         // Feed audioData to your STT service
///     }
/// }
/// ```
public class AudioCaptureManager: ObservableObject {
    private let logger = SDKLogger(category: "AudioCapture")

    private var audioEngine: AVAudioEngine?
    private var inputNode: AVAudioInputNode?

    @Published public var isRecording = false
    @Published public var audioLevel: Float = 0.0

    private let targetSampleRate = Double(RAC_STT_DEFAULT_SAMPLE_RATE)

    public init() {
        logger.info("AudioCaptureManager initialized")
    }

    /// Request microphone permission
    public func requestPermission() async -> Bool {
        #if os(iOS)
        // Use modern AVAudioApplication API for iOS 17+
        if #available(iOS 17.0, *) {
            return await AVAudioApplication.requestRecordPermission()
        } else {
            // Fallback to deprecated API for older iOS versions
            return await withCheckedContinuation { continuation in
                AVAudioSession.sharedInstance().requestRecordPermission { granted in
                    continuation.resume(returning: granted)
                }
            }
        }
        #elseif os(tvOS)
        // tvOS doesn't have AVAudioApplication, use legacy API
        return await withCheckedContinuation { continuation in
            AVAudioSession.sharedInstance().requestRecordPermission { granted in
                continuation.resume(returning: granted)
            }
        }
        #elseif os(macOS)
        // On macOS, use AVCaptureDevice for permission request
        return await withCheckedContinuation { continuation in
            AVCaptureDevice.requestAccess(for: .audio) { granted in
                continuation.resume(returning: granted)
            }
        }
        #endif
    }

    /// Start recording audio from microphone
    /// - Note: Not supported on watchOS due to AVAudioEngine limitations
    public func startRecording(onAudioData: @escaping (Data) -> Void) throws {
        guard !isRecording else {
            logger.warning("Already recording")
            return
        }

        #if os(iOS) || os(tvOS)
        let audioSession = AVAudioSession.sharedInstance()
        try audioSession.setCategory(.record, mode: .measurement)
        try audioSession.setActive(true)
        #endif

        let engine = AVAudioEngine()
        let inputNode = engine.inputNode

        #if os(macOS)
        // On macOS, Bluetooth input devices (AirPods, etc.) frequently fail to
        // start their SCO mic connection, producing silence. Detect this and
        // override to the built-in microphone before preparing the engine.
        configureMacOSInputDevice(engine: engine)
        engine.prepare()
        #endif

        let inputFormat = inputNode.outputFormat(forBus: 0)

        guard inputFormat.sampleRate > 0, inputFormat.channelCount > 0 else {
            logger.error("No valid audio input device (sampleRate=\(inputFormat.sampleRate), channels=\(inputFormat.channelCount))")
            throw AudioCaptureError.noInputDevice
        }

        logger.info("Input format: \(inputFormat.sampleRate) Hz, \(inputFormat.channelCount) channels")

        guard let outputFormat = AVAudioFormat(
            commonFormat: .pcmFormatInt16,
            sampleRate: targetSampleRate,
            channels: 1,
            interleaved: false
        ) else {
            throw AudioCaptureError.formatConversionFailed
        }

        guard let converter = AVAudioConverter(from: inputFormat, to: outputFormat) else {
            throw AudioCaptureError.formatConversionFailed
        }

        inputNode.installTap(onBus: 0, bufferSize: 4096, format: inputFormat) { [weak self] buffer, _ in
            guard let self = self else { return }

            self.updateAudioLevel(buffer: buffer)

            guard let convertedBuffer = self.convert(buffer: buffer, using: converter, to: outputFormat) else {
                return
            }

            if let audioData = self.bufferToData(buffer: convertedBuffer) {
                DispatchQueue.main.async {
                    onAudioData(audioData)
                }
            }
        }

        do {
            try engine.start()
        } catch {
            inputNode.removeTap(onBus: 0)
            throw error
        }

        self.audioEngine = engine
        self.inputNode = inputNode

        DispatchQueue.main.async {
            self.isRecording = true
        }

        logger.info("Recording started")
    }

    /// Activates the AVAudioSession without starting the audio engine.
    /// Call this to keep the app alive in the background (with UIBackgroundModes: audio)
    /// before the user is ready to record. Follow with `startRecording` when recording begins.
    public func activateAudioSession() throws {
        #if os(iOS) || os(tvOS)
        let audioSession = AVAudioSession.sharedInstance()
        try audioSession.setCategory(.record, mode: .measurement)
        try audioSession.setActive(true)
        logger.info("Audio session activated (keepalive)")
        #endif
    }

    /// Deactivates the AVAudioSession. Call this when the session is fully ended.
    public func deactivateAudioSession() {
        #if os(iOS) || os(tvOS)
        try? AVAudioSession.sharedInstance().setActive(false)
        logger.info("Audio session deactivated")
        #endif
    }

    /// Stop recording.
    /// - Parameter deactivateSession: When `true` (default) the AVAudioSession is also
    ///   deactivated. Pass `false` to keep the session alive for subsequent recordings
    ///   (e.g. between listening segments in a flow session).
    public func stopRecording(deactivateSession: Bool = true) {
        guard isRecording else { return }

        inputNode?.removeTap(onBus: 0)
        audioEngine?.stop()

        audioEngine = nil
        inputNode = nil

        #if os(iOS) || os(tvOS)
        if deactivateSession {
            try? AVAudioSession.sharedInstance().setActive(false)
        }
        #endif

        DispatchQueue.main.async {
            self.isRecording = false
            self.audioLevel = 0.0
        }

        logger.info("Recording stopped (deactivateSession=\(deactivateSession))")
    }

    // MARK: - Private Helpers

    /// Converts a PCM buffer to the target format. Internal for unit testing.
    internal func convert(
        buffer: AVAudioPCMBuffer,
        using converter: AVAudioConverter,
        to format: AVAudioFormat
    ) -> AVAudioPCMBuffer? {
        // The input block returns .endOfStream after providing one buffer.
        // On macOS the converter stays in that "finished" state across calls,
        // producing empty output for every subsequent buffer. Resetting before
        // each conversion clears the state so the next buffer is processed.
        converter.reset()

        let capacity = AVAudioFrameCount(ceil(Double(buffer.frameLength) * (format.sampleRate / buffer.format.sampleRate)))

        guard let convertedBuffer = AVAudioPCMBuffer(
            pcmFormat: format,
            frameCapacity: capacity
        ) else {
            return nil
        }

        var error: NSError?
        var hasProvidedData = false
        let inputBlock: AVAudioConverterInputBlock = { _, outStatus in
            if hasProvidedData {
                outStatus.pointee = .noDataNow
                return nil
            }
            hasProvidedData = true
            outStatus.pointee = .haveData
            return buffer
        }

        converter.convert(to: convertedBuffer, error: &error, withInputFrom: inputBlock)

        if let error = error {
            logger.error("Conversion error: \(error.localizedDescription)")
            return nil
        }

        return convertedBuffer
    }

    private func bufferToData(buffer: AVAudioPCMBuffer) -> Data? {
        guard let channelData = buffer.int16ChannelData else {
            return nil
        }

        let channelDataPointer = channelData.pointee
        let dataSize = Int(buffer.frameLength * buffer.format.streamDescription.pointee.mBytesPerFrame)

        return Data(bytes: channelDataPointer, count: dataSize)
    }

    private func updateAudioLevel(buffer: AVAudioPCMBuffer) {
        guard let channelData = buffer.floatChannelData else { return }

        let channelDataPointer = channelData.pointee
        let frames = Int(buffer.frameLength)

        // Calculate RMS (root mean square) for audio level
        var sum: Float = 0.0
        for i in 0..<frames {
            let sample = channelDataPointer[i]
            sum += sample * sample
        }

        let rms = sqrt(sum / Float(frames))
        let dbLevel = 20 * log10(rms + 0.0001) // Add small value to avoid log(0)

        // Normalize to 0-1 range (-60dB to 0dB)
        let normalizedLevel = max(0, min(1, (dbLevel + 60) / 60))

        DispatchQueue.main.async {
            self.audioLevel = normalizedLevel
        }
    }

    deinit {
        stopRecording()
    }
}

// MARK: - macOS Input Device Selection

#if os(macOS)

extension AudioCaptureManager {

    /// Detects if the default input device is Bluetooth and overrides to the
    /// built-in microphone. Bluetooth SCO mic frequently fails on macOS,
    /// producing silence. Must be called after accessing `engine.inputNode`
    /// (which creates the audio unit) but before `engine.prepare()`.
    fileprivate func configureMacOSInputDevice(engine: AVAudioEngine) {
        guard let defaultInput = MacAudioDeviceQuery.defaultInputDevice() else {
            logger.warning("Could not determine default input device")
            return
        }

        logger.info("Default input: \(defaultInput.name) (bluetooth=\(defaultInput.isBluetooth))")

        guard defaultInput.isBluetooth else { return }

        logger.warning("Default input is Bluetooth (\(defaultInput.name)) — Bluetooth SCO mic is unreliable for STT. Switching to wired input.")

        // Prefer the built-in mic
        if let builtIn = MacAudioDeviceQuery.builtInInputDevice() {
            let status = engine.setInputDevice(builtIn.deviceID)
            if status == noErr {
                logger.info("Switched input to: \(builtIn.name)")
                return
            }
            logger.error("Failed to set built-in mic (OSStatus \(status))")
        }

        // Fall back to any non-Bluetooth input device
        for device in MacAudioDeviceQuery.nonBluetoothInputDevices() {
            let status = engine.setInputDevice(device.deviceID)
            if status == noErr {
                logger.info("Switched input to: \(device.name)")
                return
            }
        }

        logger.warning("No non-Bluetooth input available — using Bluetooth as last resort")
    }
}

// MARK: - AVAudioEngine Device Override

private extension AVAudioEngine {

    /// Sets the input device for this engine via the underlying AudioUnit.
    /// Must be called after accessing `inputNode` and before `prepare()`.
    func setInputDevice(_ deviceID: AudioDeviceID) -> OSStatus {
        guard let audioUnit = inputNode.audioUnit else {
            return kAudioUnitErr_NoConnection
        }
        var mutableDeviceID = deviceID
        return AudioUnitSetProperty(
            audioUnit,
            kAudioOutputUnitProperty_CurrentDevice,
            kAudioUnitScope_Global,
            0,
            &mutableDeviceID,
            UInt32(MemoryLayout<AudioDeviceID>.size)
        )
    }
}

// MARK: - CoreAudio Device Query

/// Lightweight CoreAudio HAL queries for input device detection.
private enum MacAudioDeviceQuery {

    struct InputDeviceInfo {
        let deviceID: AudioDeviceID
        let name: String
        let transportType: UInt32

        var isBluetooth: Bool {
            transportType == kAudioDeviceTransportTypeBluetooth
                || transportType == kAudioDeviceTransportTypeBluetoothLE
        }

        var isBuiltIn: Bool {
            transportType == kAudioDeviceTransportTypeBuiltIn
        }
    }

    // MARK: Queries

    static func defaultInputDevice() -> InputDeviceInfo? {
        var deviceID = AudioDeviceID(0)
        var size = UInt32(MemoryLayout<AudioDeviceID>.size)
        var address = AudioObjectPropertyAddress(
            mSelector: kAudioHardwarePropertyDefaultInputDevice,
            mScope: kAudioObjectPropertyScopeGlobal,
            mElement: kAudioObjectPropertyElementMain
        )
        let status = AudioObjectGetPropertyData(
            AudioObjectID(kAudioObjectSystemObject),
            &address, 0, nil, &size, &deviceID
        )
        guard status == noErr, deviceID != kAudioObjectUnknown else { return nil }
        return deviceInfo(for: deviceID)
    }

    static func builtInInputDevice() -> InputDeviceInfo? {
        allInputDevices().first { $0.isBuiltIn }
    }

    static func nonBluetoothInputDevices() -> [InputDeviceInfo] {
        allInputDevices().filter { !$0.isBluetooth }
    }

    // MARK: Internal

    private static func allInputDevices() -> [InputDeviceInfo] {
        var size = UInt32(0)
        var address = AudioObjectPropertyAddress(
            mSelector: kAudioHardwarePropertyDevices,
            mScope: kAudioObjectPropertyScopeGlobal,
            mElement: kAudioObjectPropertyElementMain
        )
        guard AudioObjectGetPropertyDataSize(
            AudioObjectID(kAudioObjectSystemObject),
            &address, 0, nil, &size
        ) == noErr else { return [] }

        let count = Int(size) / MemoryLayout<AudioDeviceID>.size
        var deviceIDs = [AudioDeviceID](repeating: 0, count: count)
        guard AudioObjectGetPropertyData(
            AudioObjectID(kAudioObjectSystemObject),
            &address, 0, nil, &size, &deviceIDs
        ) == noErr else { return [] }

        return deviceIDs.compactMap { id in
            guard hasInputChannels(id) else { return nil }
            return deviceInfo(for: id)
        }
    }

    private static func deviceInfo(for deviceID: AudioDeviceID) -> InputDeviceInfo? {
        guard let name = deviceName(deviceID) else { return nil }
        return InputDeviceInfo(
            deviceID: deviceID,
            name: name,
            transportType: transportType(deviceID)
        )
    }

    private static func deviceName(_ deviceID: AudioDeviceID) -> String? {
        var name = "" as CFString
        var size = UInt32(MemoryLayout<CFString>.size)
        var address = AudioObjectPropertyAddress(
            mSelector: kAudioDevicePropertyDeviceNameCFString,
            mScope: kAudioObjectPropertyScopeGlobal,
            mElement: kAudioObjectPropertyElementMain
        )
        let status = AudioObjectGetPropertyData(deviceID, &address, 0, nil, &size, &name)
        return status == noErr ? name as String : nil
    }

    private static func transportType(_ deviceID: AudioDeviceID) -> UInt32 {
        var transport = UInt32(0)
        var size = UInt32(MemoryLayout<UInt32>.size)
        var address = AudioObjectPropertyAddress(
            mSelector: kAudioDevicePropertyTransportType,
            mScope: kAudioObjectPropertyScopeGlobal,
            mElement: kAudioObjectPropertyElementMain
        )
        AudioObjectGetPropertyData(deviceID, &address, 0, nil, &size, &transport)
        return transport
    }

    private static func hasInputChannels(_ deviceID: AudioDeviceID) -> Bool {
        var size = UInt32(0)
        var address = AudioObjectPropertyAddress(
            mSelector: kAudioDevicePropertyStreamConfiguration,
            mScope: kAudioDevicePropertyScopeInput,
            mElement: kAudioObjectPropertyElementMain
        )
        guard AudioObjectGetPropertyDataSize(deviceID, &address, 0, nil, &size) == noErr,
              size > 0 else { return false }

        let bufferListPointer = UnsafeMutableRawPointer.allocate(
            byteCount: Int(size),
            alignment: MemoryLayout<AudioBufferList>.alignment
        )
        defer { bufferListPointer.deallocate() }

        guard AudioObjectGetPropertyData(
            deviceID, &address, 0, nil, &size, bufferListPointer
        ) == noErr else { return false }

        let bufferList = bufferListPointer.assumingMemoryBound(to: AudioBufferList.self).pointee
        return bufferList.mNumberBuffers > 0 && bufferList.mBuffers.mNumberChannels > 0
    }
}

#endif

// MARK: - Errors

public enum AudioCaptureError: LocalizedError {
    case permissionDenied
    case formatConversionFailed
    case engineStartFailed
    case noInputDevice

    public var errorDescription: String? {
        switch self {
        case .permissionDenied:
            return "Microphone permission denied"
        case .formatConversionFailed:
            return "Failed to convert audio format"
        case .engineStartFailed:
            return "Failed to start audio engine"
        case .noInputDevice:
            return "No audio input device available"
        }
    }
}

File: ios/runanywhere-sdks-latest/sdk/runanywhere-flutter/packages/runanywhere/lib/features/stt/services/audio_capture_manager.dart
import 'dart:async';
import 'dart:typed_data';
import 'package:permission_handler/permission_handler.dart';
import 'package:record/record.dart';
import 'package:runanywhere/foundation/logging/sdk_logger.dart';

/// Manages audio capture from microphone for STT services.
///
/// This is a shared utility that works with any STT backend (ONNX, WhisperKit, etc.).
/// It captures audio at 16kHz mono Int16 format, which is the standard input format
/// for speech recognition models like Whisper.
///
/// Matches iOS AudioCaptureManager from Features/STT/Services/AudioCaptureManager.swift
///
/// ## Usage
/// ```dart
/// final capture = AudioCaptureManager();
/// final granted = await capture.requestPermission();
/// if (granted) {
///   await capture.startRecording((audioData) {
///     // Feed audioData to your STT service
///   });
/// }
/// ```
class AudioCaptureManager {
  final SDKLogger _logger = SDKLogger('AudioCapture');

  /// Audio recorder instance
  final AudioRecorder _recorder = AudioRecorder();

  /// Whether audio is currently being recorded
  bool _isRecording = false;

  /// Current audio level (0.0 to 1.0)
  double _audioLevel = 0.0;

  /// Target sample rate for Whisper models
  static const int targetSampleRate = 16000;

  /// Audio data callback
  void Function(Uint8List audioData)? _onAudioData;

  /// Audio stream subscription
  StreamSubscription<Uint8List>? _audioStreamSubscription;

  /// Stream controller for recording state changes
  final _recordingStateController = StreamController<bool>.broadcast();

  /// Stream controller for audio level updates
  final _audioLevelController = StreamController<double>.broadcast();

  /// Stream of recording state changes
  Stream<bool> get recordingStateStream => _recordingStateController.stream;

  /// Stream of audio level updates (0.0 to 1.0)
  Stream<double> get audioLevelStream => _audioLevelController.stream;

  /// Whether audio is currently being recorded
  bool get isRecording => _isRecording;

  /// Current audio level (0.0 to 1.0)
  double get audioLevel => _audioLevel;

  AudioCaptureManager() {
    _logger.info('AudioCaptureManager initialized');
  }

  /// Request microphone permission
  ///
  /// Returns true if permission was granted, false otherwise.
  Future<bool> requestPermission() async {
    try {
      _logger.info('Requesting microphone permission');

      // Check and request microphone permission
      final status = await Permission.microphone.request();
      if (status.isGranted) {
        _logger.info('Microphone permission granted');
        return true;
      } else if (status.isPermanentlyDenied) {
        _logger.warning(
            'Microphone permission permanently denied - user should enable in settings');
        return false;
      } else {
        _logger.warning('Microphone permission denied: $status');
        return false;
      }
    } catch (e) {
      _logger.error('Failed to request microphone permission: $e');
      return false;
    }
  }

  /// Start recording audio from microphone
  ///
  /// [onAudioData] Callback for audio data chunks (16kHz mono Int16 PCM)
  ///
  /// Throws [AudioCaptureError] if recording cannot be started.
  Future<void> startRecording(
      void Function(Uint8List audioData) onAudioData) async {
    if (_isRecording) {
      _logger.warning('Already recording');
      return;
    }

    try {
      _onAudioData = onAudioData;

      // Check if we can record
      _logger.info('Checking microphone permission...');
      if (!await _recorder.hasPermission()) {
        _logger.error('No microphone permission');
        throw AudioCaptureError.permissionDenied();
      }
      _logger.info('✅ Microphone permission granted');

      // Start streaming audio in PCM 16-bit format
      _logger.info(
          'Starting audio stream: ${targetSampleRate}Hz, mono, PCM16bits...');
      final stream = await _recorder.startStream(
        const RecordConfig(
          encoder: AudioEncoder.pcm16bits,
          sampleRate: targetSampleRate,
          numChannels: 1, // Mono
          bitRate: 256000,
        ),
      );
      _logger.info('✅ Audio stream created, setting up listener...');

      // Track data delivery
      int chunkCount = 0;

      // Listen to audio stream
      _audioStreamSubscription = stream.listen(
        (data) {
          chunkCount++;
          // Log first few chunks to verify data flow
          if (chunkCount <= 3) {
            _logger.info(
                '🎵 Audio data received: chunk #$chunkCount, ${data.length} bytes');
          }
          if (_isRecording && _onAudioData != null) {
            _onAudioData!(data);
          } else {
            _logger.warning(
                '⚠️ Audio data ignored: isRecording=$_isRecording, hasCallback=${_onAudioData != null}');
          }
        },
        onError: (Object error) {
          _logger.error('❌ Audio stream error: $error');
        },
        onDone: () {
          _logger.info('🛑 Audio stream ended (total chunks: $chunkCount)');
        },
      );

      _isRecording = true;
      _recordingStateController.add(true);

      _logger.info(
          '✅ Recording started successfully (16kHz mono PCM) - waiting for audio data...');
    } catch (e) {
      _logger.error('❌ Failed to start recording: $e');
      _onAudioData = null;
      throw AudioCaptureError.engineStartFailed();
    }
  }

  /// Stop recording
  Future<void> stopRecording() async {
    if (!_isRecording) return;

    try {
      // Cancel stream subscription
      await _audioStreamSubscription?.cancel();
      _audioStreamSubscription = null;

      // Stop the recorder
      await _recorder.stop();

      _isRecording = false;
      _audioLevel = 0.0;
      _onAudioData = null;

      _recordingStateController.add(false);
      _audioLevelController.add(0.0);

      _logger.info('Recording stopped');
    } catch (e) {
      _logger.error('Error stopping recording: $e');
    }
  }

  /// Update audio level for visualization
  ///
  /// This would be called from platform-specific audio buffer callbacks.
  /// Calculates RMS (root mean square) and normalizes to 0-1 range.
  ///
  /// [buffer] Float audio samples
  void updateAudioLevel(Float32List buffer) {
    if (buffer.isEmpty) return;

    // Calculate RMS (root mean square) for audio level
    double sum = 0.0;
    for (final sample in buffer) {
      sum += sample * sample;
    }

    final rms = _sqrt(sum / buffer.length);
    final dbLevel =
        20 * _log10(rms + 0.0001); // Add small value to avoid log(0)

    // Normalize to 0-1 range (-60dB to 0dB)
    final normalizedLevel = (dbLevel + 60) / 60;
    _audioLevel = normalizedLevel.clamp(0.0, 1.0);

    _audioLevelController.add(_audioLevel);
  }

  /// Convert audio buffer to PCM data
  ///
  /// This would be called from platform-specific audio buffer callbacks
  /// to convert audio samples to the format expected by STT models.
  ///
  /// [buffer] Int16 audio samples (16-bit PCM)
  /// Returns PCM data as bytes
  Uint8List bufferToData(Int16List buffer) {
    // Convert Int16 samples to bytes (little-endian)
    final bytes = BytesBuilder();
    for (final sample in buffer) {
      bytes.add([
        sample & 0xFF, // Low byte
        (sample >> 8) & 0xFF, // High byte
      ]);
    }
    return bytes.toBytes();
  }

  /// Called when audio data is available from the platform
  ///
  /// This would be called from platform-specific audio buffer callbacks.
  ///
  /// [audioData] Raw PCM audio data (16kHz mono Int16)
  void onAudioDataAvailable(Uint8List audioData) {
    final callback = _onAudioData;
    if (callback != null && _isRecording) {
      callback(audioData);
    }
  }

  /// Dispose resources
  void dispose() {
    unawaited(stopRecording());
    unawaited(_recordingStateController.close());
    unawaited(_audioLevelController.close());
  }

  // MARK: - Private Math Helpers

  /// Square root helper
  double _sqrt(double x) {
    if (x <= 0) return 0.0;
    // Use Newton's method for square root approximation
    double guess = x / 2;
    for (int i = 0; i < 10; i++) {
      guess = (guess + x / guess) / 2;
    }
    return guess;
  }

  /// Base-10 logarithm helper
  double _log10(double x) {
    if (x <= 0) return -60.0; // Return minimum dB level
    // Use natural log and convert to log10
    // log10(x) = ln(x) / ln(10)
    return _ln(x) / 2.302585092994046; // ln(10)
  }

  /// Natural logarithm helper (using Taylor series)
  double _ln(double x) {
    if (x <= 0) return double.negativeInfinity;
    if (x == 1) return 0.0;

    // For better convergence, use ln(x) = 2 * atanh((x-1)/(x+1))
    final y = (x - 1) / (x + 1);
    double result = 0.0;
    double term = y;
    const maxIterations = 20;

    for (int i = 0; i < maxIterations; i++) {
      result += term / (2 * i + 1);
      term *= y * y;
    }

    return 2 * result;
  }
}

// MARK: - Errors

/// Audio capture errors
/// Matches iOS AudioCaptureError from AudioCaptureManager.swift
class AudioCaptureError implements Exception {
  final String message;
  final AudioCaptureErrorType type;

  AudioCaptureError._(this.message, this.type);

  factory AudioCaptureError.permissionDenied() {
    return AudioCaptureError._(
      'Microphone permission denied',
      AudioCaptureErrorType.permissionDenied,
    );
  }

  factory AudioCaptureError.formatConversionFailed() {
    return AudioCaptureError._(
      'Failed to convert audio format',
      AudioCaptureErrorType.formatConversionFailed,
    );
  }

  factory AudioCaptureError.engineStartFailed() {
    return AudioCaptureError._(
      'Failed to start audio engine',
      AudioCaptureErrorType.engineStartFailed,
    );
  }

  @override
  String toString() => 'AudioCaptureError: $message';
}

/// Audio capture error types
enum AudioCaptureErrorType {
  permissionDenied,
  formatConversionFailed,
  engineStartFailed,
}
